wasm/src/main/scala/ir2wasm/HelperFunctions.scala

```diff
@@ -784,4 +787,38 @@ object HelperFunctions {
     fctx.buildAndAddToContext()
   }
 
+  def genCloneInstance(clazz: LinkedClass)(implicit ctx: WasmContext): Unit = {
+    import WasmImmediate._
+    val info = ctx.getClassInfo(clazz.name.name)
+    if (info.interfaces.contains(IRNames.CloneableClass)) {
+      val heapType =
+        WasmHeapType.Type(WasmTypeName.WasmStructTypeName(clazz.name.name))
+      val fctx = WasmFunctionContext(
+        Names.WasmFunctionName.clone(clazz.name.name),
+        List("from" -> WasmRefType(WasmHeapType.ObjectType)),
+        List(WasmRefType(WasmHeapType.ObjectType))
+      )
+      val List(fromParam) = fctx.paramIndices
+      import fctx.instrs
+
+      val result = fctx.addLocal(fctx.genSyntheticLocalName(), WasmRefNullType(heapType))
+
+      instrs += CALL(FuncIdx(WasmFunctionName.newDefault(clazz.name.name)))
+      instrs += LOCAL_SET(result)
+      info.allFieldDefs.foreach { field =>
+        val fieldIdx = info.getFieldIdx(field.name.name)
+        instrs += LOCAL_GET(result)
+        instrs += LOCAL_GET(fromParam)
+        instrs += REF_CAST(HeapType(heapType))
+        instrs += STRUCT_GET(TypeIdx(WasmTypeName.WasmStructTypeName(clazz.name.name)), fieldIdx)
+        instrs += STRUCT_SET(TypeIdx(WasmTypeName.WasmStructTypeName(clazz.name.name)), fieldIdx)
+      }
+      instrs += LOCAL_GET(result)
+      instrs += REF_AS_NOT_NULL
+      // instrs += REF_CAST(HeapType(WasmHeapType.ObjectType))
+      fctx.buildAndAddToContext(WasmFunctionType.cloneFunction)
```

Comment:
Is there a reason that you could not do the `buildAndAddToContext()` without argument?

Reply:
because otherwise, we need to store the name in context and pass it around to the caller of the clone function like this:
https://github.com/tanishiking/scala-wasm/pull/36/commits/27f397805fa87f3e906142948cf78267ad34c58e

`(type $fun_type___1___ty (func (param anyref) (result anyref)))` and `(type $clone_function_type (func (param anyref) (result anyref)))` are not substitutable, it seems.